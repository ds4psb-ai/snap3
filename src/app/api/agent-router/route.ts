import { NextRequest, NextResponse } from 'next/server';
import { agentRouter } from '../../../ai-collab/agent-router';

export async function POST(request: NextRequest) {
  try {
    const { fileChanges, commitMessage, correlationId } = await request.json();
    
    const context = agentRouter.detectContext(fileChanges || [], commitMessage);
    const decision = await agentRouter.routeDecision(context, correlationId || `AR-${Date.now()}`);
    
    return NextResponse.json(decision);
  } catch (error) {
    return NextResponse.json({
      error: 'Agent routing failed',
      details: error instanceof Error ? error.message : 'Unknown error'
    }, { status: 500 });
  }
}