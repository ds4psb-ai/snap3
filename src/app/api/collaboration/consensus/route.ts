import { NextRequest, NextResponse } from 'next/server';

interface ConsensusResult {
  score: number;
  status: 'pending' | 'agreed' | 'disagreed' | 'error';
  lastUpdate: string;
  participants: string[];
}

export async function GET(request: NextRequest) {
  try {
    // 실제로는 합의 시스템에서 가져옴
    const consensusStatus: ConsensusResult = {
      score: 0.91,
      status: 'agreed',
      lastUpdate: new Date().toISOString(),
      participants: ['GPT-5 Pro', 'ClaudeCode', 'Cursor']
    };

    return NextResponse.json({
      consensus: consensusStatus,
      timestamp: new Date().toISOString()
    });

  } catch (error) {
    console.error('Consensus API error:', error);
    return NextResponse.json(
      { error: 'Failed to fetch consensus status' },
      { status: 500 }
    );
  }
}
