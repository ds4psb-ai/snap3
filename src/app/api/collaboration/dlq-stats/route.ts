import { NextRequest, NextResponse } from 'next/server';

interface DLQStats {
  failedJobs: number;
  errorTypes: Record<string, number>;
  lastError: string;
  recoveryRate: number;
}

export async function GET(request: NextRequest) {
  try {
    // 실제로는 DLQ 시스템에서 가져옴
    const dlqStats: DLQStats = {
      failedJobs: 2,
      errorTypes: {
        'timeout': 1,
        'validation_error': 1,
        'network_error': 0
      },
      lastError: '2025-08-20T20:25:00Z',
      recoveryRate: 0.85
    };

    return NextResponse.json({
      dlq: dlqStats,
      timestamp: new Date().toISOString()
    });

  } catch (error) {
    console.error('DLQ stats API error:', error);
    return NextResponse.json(
      { error: 'Failed to fetch DLQ stats' },
      { status: 500 }
    );
  }
}
