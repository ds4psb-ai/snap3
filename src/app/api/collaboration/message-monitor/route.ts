import { NextRequest, NextResponse } from 'next/server';
import { promises as fs } from 'fs';
import path from 'path';

interface MessageInfo {
  filename: string;
  content: string;
  timestamp: string;
  priority: 'critical' | 'high' | 'normal';
  processed: boolean;
}

interface MonitorResponse {
  status: 'success' | 'error';
  messages: MessageInfo[];
  lastCheck: string;
  hasNewMessages: boolean;
}

// 메시지 우선순위 분석
function analyzePriority(content: string): 'critical' | 'high' | 'normal' {
  const lowerContent = content.toLowerCase();
  
  if (lowerContent.includes('critical') || lowerContent.includes('🚨')) {
    return 'critical';
  }
  
  if (lowerContent.includes('high') || lowerContent.includes('⚠️')) {
    return 'high';
  }
  
  return 'normal';
}

// 메시지 파일 읽기
async function readMessageFiles(): Promise<MessageInfo[]> {
  const repoRoot = process.cwd();
  const messages: MessageInfo[] = [];
  
  try {
    const files = await fs.readdir(repoRoot);
    const messageFiles = files.filter(file => file.startsWith('.collab-msg-'));
    
    for (const filename of messageFiles) {
      try {
        const filePath = path.join(repoRoot, filename);
        const content = await fs.readFile(filePath, 'utf-8');
        const stats = await fs.stat(filePath);
        
        messages.push({
          filename,
          content: content.trim(),
          timestamp: stats.mtime.toISOString(),
          priority: analyzePriority(content),
          processed: false
        });
      } catch (error) {
        console.error(`Error reading message file ${filename}:`, error);
      }
    }
    
    // 타임스탬프 기준으로 정렬 (최신순)
    messages.sort((a, b) => new Date(b.timestamp).getTime() - new Date(a.timestamp).getTime());
    
  } catch (error) {
    console.error('Error reading message directory:', error);
  }
  
  return messages;
}

// 메시지 처리 상태 업데이트
async function markMessageProcessed(filename: string): Promise<void> {
  const repoRoot = process.cwd();
  const originalPath = path.join(repoRoot, filename);
  const processedPath = path.join(repoRoot, `${filename}.processed`);
  
  try {
    await fs.rename(originalPath, processedPath);
    console.log(`Message processed: ${filename} → ${filename}.processed`);
  } catch (error) {
    console.error(`Error marking message as processed: ${filename}`, error);
  }
}

export async function GET(request: NextRequest): Promise<NextResponse<MonitorResponse>> {
  try {
    const { searchParams } = new URL(request.url);
    const action = searchParams.get('action');
    
    if (action === 'mark-processed') {
      const filename = searchParams.get('filename');
      if (filename) {
        await markMessageProcessed(filename);
        return NextResponse.json({ 
          status: 'success', 
          message: `Message ${filename} marked as processed` 
        });
      }
    }
    
    // 메시지 파일 읽기
    const messages = await readMessageFiles();
    const hasNewMessages = messages.length > 0;
    
    const response: MonitorResponse = {
      status: 'success',
      messages,
      lastCheck: new Date().toISOString(),
      hasNewMessages
    };
    
    return NextResponse.json(response);
    
  } catch (error) {
    console.error('Message monitor error:', error);
    return NextResponse.json(
      { 
        status: 'error', 
        error: 'Failed to monitor messages',
        messages: [],
        lastCheck: new Date().toISOString(),
        hasNewMessages: false
      },
      { status: 500 }
    );
  }
}

export async function POST(request: NextRequest): Promise<NextResponse> {
  try {
    const body = await request.json();
    const { action, filename } = body;
    
    if (action === 'mark-processed' && filename) {
      await markMessageProcessed(filename);
      return NextResponse.json({ 
        status: 'success', 
        message: `Message ${filename} marked as processed` 
      });
    }
    
    return NextResponse.json(
      { status: 'error', error: 'Invalid action' },
      { status: 400 }
    );
    
  } catch (error) {
    console.error('Message monitor POST error:', error);
    return NextResponse.json(
      { status: 'error', error: 'Failed to process request' },
      { status: 500 }
    );
  }
}
