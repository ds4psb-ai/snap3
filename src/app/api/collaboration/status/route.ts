import { NextRequest, NextResponse } from 'next/server';

interface TerminalState {
  id: 'T1' | 'T2' | 'T3' | 'T4';
  name: string;
  port: number;
  status: 'online' | 'offline' | 'error' | 'starting';
  lastHeartbeat: string;
  load: number;
  memory: number;
}

export async function GET(request: NextRequest) {
  try {
    // 실제로는 각 터미널의 헬스체크 API를 호출
    const terminalStates: TerminalState[] = [
      {
        id: 'T1',
        name: 'Main Terminal',
        port: 3000,
        status: 'online',
        lastHeartbeat: new Date().toISOString(),
        load: 45,
        memory: 68
      },
      {
        id: 'T2',
        name: 'Jobs Terminal',
        port: 8080,
        status: 'online',
        lastHeartbeat: new Date(Date.now() - 5000).toISOString(),
        load: 72,
        memory: 85
      },
      {
        id: 'T3',
        name: 'T2VDP Terminal',
        port: 8082,
        status: 'starting',
        lastHeartbeat: new Date(Date.now() - 15000).toISOString(),
        load: 0,
        memory: 0
      },
      {
        id: 'T4',
        name: 'Storage Terminal',
        port: 8083,
        status: 'online',
        lastHeartbeat: new Date(Date.now() - 2000).toISOString(),
        load: 23,
        memory: 42
      }
    ];

    return NextResponse.json({
      terminals: terminalStates,
      timestamp: new Date().toISOString()
    });

  } catch (error) {
    console.error('Terminal status API error:', error);
    return NextResponse.json(
      { error: 'Failed to fetch terminal status' },
      { status: 500 }
    );
  }
}
