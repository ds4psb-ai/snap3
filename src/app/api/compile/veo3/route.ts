import { NextRequest, NextResponse } from 'next/server';
import { z } from 'zod';
import { Problems } from '@/lib/errors/problem';
import { VEO3_PROMPT_SCHEMA } from '@/lib/schemas/veo3.zod';

export async function POST(request: NextRequest) {
  const instance = '/api/compile/veo3';
  
  try {
    const body = await request.json();
    const validatedData = VEO3_PROMPT_SCHEMA.parse(body);
    
    // Validate constraints
    if (validatedData.duration !== 8) {
      return Problems.invalidDuration(validatedData.duration, instance);
    }
    
    if (validatedData.aspect !== '16:9') {
      return Problems.unsupportedAspectRatio(validatedData.aspect, instance);
    }
    
    // TODO: Implement Veo3 prompt compilation
    return NextResponse.json({
      id: 'veo3-' + Date.now(),
      prompt: validatedData.prompt,
      duration: validatedData.duration,
      aspect: validatedData.aspect,
      resolution: validatedData.resolution,
      status: 'compiled',
    });
  } catch (error) {
    if (error instanceof z.ZodError) {
      const violations = error.errors.map(err => ({
        field: err.path.join('.'),
        message: err.message,
        code: 'VALIDATION_ERROR',
      }));
      return Problems.validation(violations, instance);
    }
    
    return Problems.validation([{
      field: 'request',
      message: 'Invalid Veo3 prompt format',
      code: 'INVALID_REQUEST',
    }], instance);
  }
}

export async function GET() {
  return Problems.methodNotAllowed('GET', ['POST'], '/api/compile/veo3');
}




