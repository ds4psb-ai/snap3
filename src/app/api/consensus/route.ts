import { NextRequest, NextResponse } from 'next/server';

interface AgentProposal {
  id: string;
  agent: 'gpt5' | 'claude' | 'cursor';
  agentName: string;
  proposal: string;
  reasoning: string;
  confidence: number;
  risks: string[];
  limitations: string[];
  timestamp: string;
  status: 'pending' | 'approved' | 'rejected';
}

interface ConsensusResult {
  decision: string;
  risks: string[];
  nextActions: string[];
  rollbackPlan: string;
  confidence: number;
  slaStatus: 'good' | 'warning' | 'error';
  sloBudget: number;
}

interface SLOMetrics {
  p95ResponseTime: number;
  successRate: number;
  errorBudget: number;
  deploymentFrequency: number;
  leadTime: number;
  changeFailureRate: number;
  mttr: number;
}

export async function GET(request: NextRequest) {
  try {
    // 실제로는 데이터베이스나 외부 API에서 가져옴
    const proposals: AgentProposal[] = [
      {
        id: '1',
        agent: 'gpt5',
        agentName: 'GPT-5 Pro (전략)',
        proposal: 'LangGraph + Temporal 기반 오케스트레이션 시스템 즉시 구현',
        reasoning: '현재 VDP 파이프라인의 안정성과 확장성을 근본적으로 개선할 수 있는 가장 효과적인 방법입니다.',
        confidence: 0.95,
        risks: ['구현 복잡도 증가', '학습 곡선', '초기 설정 시간'],
        limitations: ['기존 시스템과의 통합 필요', '팀 교육 시간 필요'],
        timestamp: new Date().toISOString(),
        status: 'approved'
      },
      {
        id: '2',
        agent: 'claude',
        agentName: 'ClaudeCode (백엔드)',
        proposal: 'Circuit Breaker + DLQ + SAGA 패턴 우선 구현',
        reasoning: '신뢰성 패턴을 먼저 구현하여 시스템 안정성을 확보한 후, 점진적으로 고급 기능을 추가하는 것이 안전합니다.',
        confidence: 0.88,
        risks: ['기존 코드 수정 필요', '테스트 커버리지 확보'],
        limitations: ['점진적 구현으로 인한 일시적 복잡도'],
        timestamp: new Date(Date.now() - 300000).toISOString(),
        status: 'approved'
      },
      {
        id: '3',
        agent: 'cursor',
        agentName: 'Cursor (UX)',
        proposal: 'People+AI 가이드 준수하는 합의 콘솔 UI 구현',
        reasoning: '사용자가 AI 시스템의 의사결정 과정을 이해하고 제어할 수 있도록 투명성과 설명가능성을 제공해야 합니다.',
        confidence: 0.92,
        risks: ['UI 복잡도 증가', '성능 오버헤드'],
        limitations: ['실시간 업데이트로 인한 리소스 사용'],
        timestamp: new Date(Date.now() - 600000).toISOString(),
        status: 'approved'
      }
    ];

    const consensus: ConsensusResult = {
      decision: 'Phase-1: LangGraph + Temporal + 신뢰성 패턴 병렬 구현',
      risks: [
        '구현 복잡도 증가로 인한 초기 지연',
        '기존 시스템과의 통합 리스크',
        '팀 학습 곡선'
      ],
      nextActions: [
        'LangGraph 상태그래프 골격 구현 (ClaudeCode)',
        'Temporal 워커 보일러플레이트 생성 (ClaudeCode)',
        'Circuit Breaker + DLQ 패턴 구현 (ClaudeCode)',
        '합의 콘솔 UI 완성 (Cursor)',
        'SLO/SLA 모니터링 대시보드 구현 (Cursor)'
      ],
      rollbackPlan: '기존 시스템으로 즉시 롤백, 점진적 재구현',
      confidence: 0.91,
      slaStatus: 'good',
      sloBudget: 85
    };

    const sloMetrics: SLOMetrics = {
      p95ResponseTime: 274,
      successRate: 99.7,
      errorBudget: 85,
      deploymentFrequency: 12,
      leadTime: 45,
      changeFailureRate: 2.3,
      mttr: 15
    };

    return NextResponse.json({
      proposals,
      consensus,
      sloMetrics,
      timestamp: new Date().toISOString()
    });

  } catch (error) {
    console.error('Consensus API error:', error);
    return NextResponse.json(
      { error: 'Failed to fetch consensus data' },
      { status: 500 }
    );
  }
}

export async function POST(request: NextRequest) {
  try {
    const { action, data } = await request.json();

    switch (action) {
      case 'rollback':
        // 실제 롤백 로직 구현
        console.log('Rollback initiated:', data);
        return NextResponse.json({ 
          success: true, 
          message: 'Rollback initiated successfully',
          timestamp: new Date().toISOString()
        });

      case 'approve':
        // 승인 로직 구현
        console.log('Proposal approved:', data);
        return NextResponse.json({ 
          success: true, 
          message: 'Proposal approved successfully',
          timestamp: new Date().toISOString()
        });

      case 'reject':
        // 거부 로직 구현
        console.log('Proposal rejected:', data);
        return NextResponse.json({ 
          success: true, 
          message: 'Proposal rejected successfully',
          timestamp: new Date().toISOString()
        });

      default:
        return NextResponse.json(
          { error: 'Invalid action' },
          { status: 400 }
        );
    }

  } catch (error) {
    console.error('Consensus API POST error:', error);
    return NextResponse.json(
      { error: 'Failed to process consensus action' },
      { status: 500 }
    );
  }
}
