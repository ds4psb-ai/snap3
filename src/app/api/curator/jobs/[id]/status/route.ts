import { NextRequest, NextResponse } from 'next/server';

// Import jobs map from create route (in production, use shared storage)
let jobs: Map<string, any>;
try {
  // This is a workaround for development - in production, use Redis/Database
  jobs = new Map();
} catch (error) {
  jobs = new Map();
}

export async function GET(
  request: NextRequest,
  { params }: { params: { id: string } }
) {
  try {
    const jobId = params.id;

    if (!jobId) {
      return NextResponse.json(
        {
          type: 'https://snap3.example/problems/validation-error',
          title: 'Validation Error',
          status: 400,
          detail: 'Job ID is required',
          instance: request.url,
        },
        { status: 400 }
      );
    }

    // In production, fetch from Redis/Database
    // For now, simulate job status
    const job = jobs.get(jobId) || generateMockJob(jobId);

    if (!job) {
      return NextResponse.json(
        {
          type: 'https://snap3.example/problems/not-found',
          title: 'Job Not Found',
          status: 404,
          detail: `Job with ID ${jobId} not found`,
          instance: request.url,
        },
        { status: 404 }
      );
    }

    return NextResponse.json({
      jobId: job.id,
      uploadId: job.uploadId,
      status: job.status,
      progress: job.progress,
      fileName: job.fileName,
      fileType: job.fileType,
      gcsPath: job.gcsPath,
      createdAt: job.createdAt,
      updatedAt: job.updatedAt,
      error: job.error,
    });

  } catch (error) {
    console.error('Job status fetch error:', error);
    
    return NextResponse.json(
      {
        type: 'https://snap3.example/problems/internal-error',
        title: 'Internal Server Error',
        status: 500,
        detail: 'Failed to fetch job status',
        instance: request.url,
        timestamp: new Date().toISOString(),
      },
      { status: 500 }
    );
  }
}

// Generate mock job for demonstration
function generateMockJob(jobId: string) {
  const now = new Date();
  const createdAt = new Date(now.getTime() - Math.random() * 60000); // Random time in last minute
  
  // Simulate job progression over time
  const elapsed = now.getTime() - createdAt.getTime();
  let status: 'pending' | 'processing' | 'completed' | 'failed' = 'pending';
  let progress = 0;

  if (elapsed > 2000) {
    status = 'processing';
    progress = Math.min(90, Math.floor((elapsed - 2000) / 100));
  }
  
  if (elapsed > 10000) {
    status = 'completed';
    progress = 100;
  }

  return {
    id: jobId,
    uploadId: `upload_${jobId}`,
    gcsPath: `uploads/mock-${jobId}.csv`,
    fileName: `file-${jobId}.csv`,
    fileType: 'csv',
    status,
    progress,
    createdAt: createdAt.toISOString(),
    updatedAt: now.toISOString(),
  };
}