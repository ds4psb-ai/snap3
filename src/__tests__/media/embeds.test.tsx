/**
 * Comprehensive embed security tests
 * Validates embed policy enforcement and security headers
 */

import { describe, expect, test, beforeEach } from '@jest/globals';
import '@testing-library/jest-dom';
import { NextRequest } from 'next/server';
import { POST } from '../../app/api/embed-meta/route';
import { render } from '@testing-library/react';
import { EmbedFrame, SecureEmbed } from '../../components/EmbedFrame';

// Mock console.warn to capture security warnings
const consoleMock = {
  warn: jest.fn()
};
global.console = { ...global.console, ...consoleMock };

describe('Embed Security Policy', () => {
  beforeEach(() => {
    consoleMock.warn.mockClear();
  });

  describe('API Route: /embed-meta', () => {
    describe('✅ Allowed Domains', () => {
      test('should accept YouTube embed URLs', async () => {
        const request = new NextRequest('http://localhost:3000/api/embed-meta', {
          method: 'POST',
          body: JSON.stringify({
            url: 'https://www.youtube.com/embed/dQw4w9WgXcQ'
          }),
          headers: { 'Content-Type': 'application/json' }
        });

        const response = await POST(request);
        const data = await response.json();

        expect(response.status).toBe(200);
        expect(data.provider).toBe('youtube');
        expect(data.embedUrl).toMatch(/^https:\/\/www\.youtube\.com\/embed\/[A-Za-z0-9_-]+/);
        expect(data.videoId).toBe('dQw4w9WgXcQ');
      });

      test('should accept Vimeo embed URLs', async () => {
        const request = new NextRequest('http://localhost:3000/api/embed-meta', {
          method: 'POST',
          body: JSON.stringify({
            url: 'https://player.vimeo.com/video/123456789'
          }),
          headers: { 'Content-Type': 'application/json' }
        });

        const response = await POST(request);
        const data = await response.json();

        expect(response.status).toBe(200);
        expect(data.provider).toBe('vimeo');
        expect(data.embedUrl).toMatch(/^https:\/\/player\.vimeo\.com\/video\/\d+/);
        expect(data.videoId).toBe('123456789');
      });

      test('should normalize YouTube watch URLs to embed URLs', async () => {
        const request = new NextRequest('http://localhost:3000/api/embed-meta', {
          method: 'POST',
          body: JSON.stringify({
            url: 'https://www.youtube.com/watch?v=dQw4w9WgXcQ'
          }),
          headers: { 'Content-Type': 'application/json' }
        });

        const response = await POST(request);
        const data = await response.json();

        expect(response.status).toBe(200);
        expect(data.embedUrl).toBe('https://www.youtube.com/embed/dQw4w9WgXcQ');
        expect(data.originalUrl).toBe('https://www.youtube.com/watch?v=dQw4w9WgXcQ');
      });
    });

    describe('❌ Forbidden Domains', () => {
      const forbiddenUrls = [
        { url: 'https://example.com/video.mp4', expectedStatus: 403 },
        { url: 'https://malicious-site.com/embed/123', expectedStatus: 403 },
        { url: 'https://youtube-proxy.com/embed/123', expectedStatus: 403 },
        { url: 'https://dailymotion.com/embed/xyz', expectedStatus: 403 },
        { url: 'https://twitch.tv/embed/channel', expectedStatus: 403 },
        { url: 'https://badsite.com/vimeo.com/123456789', expectedStatus: 403 }, // Not actually vimeo.com
        { url: './assets/video.mp4', expectedStatus: 400 }, // Invalid URL format
        { url: '/videos/sample.mp4', expectedStatus: 400 } // Invalid URL format
      ];

      test.each(forbiddenUrls)('should reject unauthorized domain: $url', async ({ url, expectedStatus }) => {
        const request = new NextRequest('http://localhost:3000/api/embed-meta', {
          method: 'POST',
          body: JSON.stringify({ url }),
          headers: { 'Content-Type': 'application/json' }
        });

        const response = await POST(request);
        const data = await response.json();

        // Should return Problem+JSON format
        expect(response.status).toBe(expectedStatus);
        expect(response.headers.get('Content-Type')).toBe('application/problem+json');
        
        if (expectedStatus === 403) {
          // Validate EMBED_DENIED Problem+JSON structure
          expect(data).toHaveProperty('type');
          expect(data).toHaveProperty('title', 'Embed denied');
          expect(data).toHaveProperty('status', 403);
          expect(data).toHaveProperty('code', 'EMBED_DENIED');
          expect(data).toHaveProperty('fix');
          expect(data.detail).toContain('YouTube and Vimeo embeds');
        } else {
          // Validate VALIDATION_ERROR Problem+JSON structure
          expect(data).toHaveProperty('code', 'VALIDATION_ERROR');
          expect(data).toHaveProperty('violations');
          expect(data.violations[0].field).toBe('url');
        }
      });
    });

    describe('🔍 Validation Errors', () => {
      test('should reject invalid URL format', async () => {
        const request = new NextRequest('http://localhost:3000/api/embed-meta', {
          method: 'POST',
          body: JSON.stringify({
            url: 'not-a-valid-url'
          }),
          headers: { 'Content-Type': 'application/json' }
        });

        const response = await POST(request);
        const data = await response.json();

        expect(response.status).toBe(400);
        expect(response.headers.get('Content-Type')).toBe('application/problem+json');
        expect(data.code).toBe('VALIDATION_ERROR');
        expect(data.violations).toHaveLength(1);
        expect(data.violations[0].field).toBe('url');
      });

      test('should reject missing URL field', async () => {
        const request = new NextRequest('http://localhost:3000/api/embed-meta', {
          method: 'POST',
          body: JSON.stringify({}),
          headers: { 'Content-Type': 'application/json' }
        });

        const response = await POST(request);
        const data = await response.json();

        expect(response.status).toBe(400);
        expect(data.code).toBe('VALIDATION_ERROR');
        expect(data.violations[0].field).toBe('url');
      });
    });
  });

  describe('Component: EmbedFrame', () => {
    describe('✅ Security Validation', () => {
      test('should render YouTube embeds', () => {
        const { container } = render(
          <EmbedFrame src="https://www.youtube.com/embed/dQw4w9WgXcQ" />
        );

        const iframe = container.querySelector('iframe');
        expect(iframe).toBeInTheDocument();
        expect(iframe?.getAttribute('src')).toBe('https://www.youtube.com/embed/dQw4w9WgXcQ');
      });

      test('should render Vimeo embeds', () => {
        const { container } = render(
          <EmbedFrame src="https://player.vimeo.com/video/123456789" />
        );

        const iframe = container.querySelector('iframe');
        expect(iframe).toBeInTheDocument();
        expect(iframe?.getAttribute('src')).toBe('https://player.vimeo.com/video/123456789');
      });

      test('should block unauthorized domains', () => {
        const { container } = render(
          <EmbedFrame src="https://malicious-site.com/embed/123" />
        );

        const iframe = container.querySelector('iframe');
        expect(iframe).not.toBeInTheDocument();
        
        const errorDiv = container.querySelector('.embed-frame-error');
        expect(errorDiv).toBeInTheDocument();
        expect(errorDiv?.textContent).toContain('Embed Blocked');
        
        expect(consoleMock.warn).toHaveBeenCalledWith(
          'EmbedFrame: Unauthorized embed source blocked:',
          'https://malicious-site.com/embed/123'
        );
      });
    });

    describe('🛡️ Security Attributes', () => {
      test('should have proper security attributes', () => {
        const { container } = render(
          <EmbedFrame src="https://www.youtube.com/embed/test123" />
        );

        const iframe = container.querySelector('iframe');
        expect(iframe).toHaveAttribute('referrerPolicy', 'no-referrer');
        expect(iframe).toHaveAttribute('sandbox', 'allow-scripts allow-same-origin allow-presentation allow-popups');
        expect(iframe).toHaveAttribute('allow', 'autoplay; fullscreen; picture-in-picture');
        expect(iframe).toHaveAttribute('frameBorder', '0');
        expect(iframe).toHaveAttribute('allowFullScreen');
        expect(iframe).toHaveAttribute('credentialless', 'true');
      });

      test('should have default accessibility attributes', () => {
        const { container } = render(
          <EmbedFrame src="https://www.youtube.com/embed/test123" />
        );

        const iframe = container.querySelector('iframe');
        expect(iframe).toHaveAttribute('title', 'Embedded Video');
        expect(iframe).toHaveAttribute('loading', 'lazy');
      });

      test('should accept custom title and loading attributes', () => {
        const { container } = render(
          <EmbedFrame 
            src="https://www.youtube.com/embed/test123"
            title="Custom Video Title"
            loading="eager"
          />
        );

        const iframe = container.querySelector('iframe');
        expect(iframe).toHaveAttribute('title', 'Custom Video Title');
        expect(iframe).toHaveAttribute('loading', 'eager');
      });
    });

    describe('🔧 SecureEmbed HOC', () => {
      test('should render successfully with valid embed', () => {
        const { container } = render(
          <SecureEmbed src="https://www.youtube.com/embed/test123" />
        );

        const iframe = container.querySelector('iframe');
        expect(iframe).toBeInTheDocument();
      });

      test('should show custom fallback for invalid embed', () => {
        const fallback = <div data-testid="custom-fallback">Custom Error</div>;
        
        const { getByTestId } = render(
          <SecureEmbed 
            src="https://malicious-site.com/embed/123"
            fallback={fallback}
          />
        );

        expect(getByTestId('custom-fallback')).toBeInTheDocument();
      });

      test('should show default fallback when no custom fallback provided', () => {
        const { container } = render(
          <SecureEmbed src="https://malicious-site.com/embed/123" />
        );

        const fallback = container.querySelector('.embed-error-fallback');
        expect(fallback).toBeInTheDocument();
        expect(fallback?.textContent).toContain('Unable to load embed');
      });
    });
  });

  describe('🔒 CSP Header Validation', () => {
    // Note: This would typically be tested in E2E tests with actual headers
    // Here we validate the expected header structure
    test('should define proper frame-src policy', () => {
      const expectedFrameSrc = "frame-src 'self' https://www.youtube.com https://youtube.com https://player.vimeo.com";
      const expectedFrameAncestors = "frame-ancestors 'self'";
      
      // These would be tested against actual Next.js config in integration tests
      expect(expectedFrameSrc).toContain('youtube.com');
      expect(expectedFrameSrc).toContain('player.vimeo.com');
      expect(expectedFrameAncestors).toBe("frame-ancestors 'self'");
    });

    test('should define restrictive permissions policy', () => {
      const expectedPermissions = [
        'autoplay=(self)',
        'camera=()',
        'microphone=()',
        'geolocation=()'
      ];
      
      expectedPermissions.forEach(permission => {
        expect(permission).toMatch(/^[a-z-]+=/);
      });
    });
  });

  describe('📊 Component Snapshots', () => {
    test('iframe properties snapshot for YouTube', () => {
      const { container } = render(
        <EmbedFrame 
          src="https://www.youtube.com/embed/test123"
          width={560}
          height={315}
        />
      );

      const iframe = container.querySelector('iframe');
      const attributes = {
        src: iframe?.getAttribute('src'),
        width: iframe?.getAttribute('width'),
        height: iframe?.getAttribute('height'),
        referrerPolicy: iframe?.getAttribute('referrerPolicy'),
        sandbox: iframe?.getAttribute('sandbox'),
        allow: iframe?.getAttribute('allow'),
        frameBorder: iframe?.getAttribute('frameBorder'),
        allowFullScreen: iframe?.hasAttribute('allowFullScreen'),
        credentialless: iframe?.getAttribute('credentialless')
      };

      expect(attributes).toMatchSnapshot({
        src: 'https://www.youtube.com/embed/test123',
        width: '560',
        height: '315',
        referrerPolicy: 'no-referrer',
        sandbox: 'allow-scripts allow-same-origin allow-presentation allow-popups',
        allow: 'autoplay; fullscreen; picture-in-picture',
        frameBorder: '0',
        allowFullScreen: true,
        credentialless: 'true'
      });
    });
  });
});