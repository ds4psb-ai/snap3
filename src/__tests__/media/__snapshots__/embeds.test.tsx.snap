// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Embed Security Policy ðŸ“Š Component Snapshots iframe properties snapshot for YouTube 1`] = `
{
  "allow": "autoplay; fullscreen; picture-in-picture",
  "allowFullScreen": true,
  "credentialless": "true",
  "frameBorder": "0",
  "height": "315",
  "referrerPolicy": "no-referrer",
  "sandbox": "allow-scripts allow-same-origin allow-presentation allow-popups",
  "src": "https://www.youtube.com/embed/test123",
  "width": "560",
}
`;
